Matching Algorithm implemented & Dynamic Results shows

OliviaY1 opened the pull request:
Matching Algorithm based on tutor rating & student rating.
SearchCourseResultView shows the number of results & course code that is searched.

OliviaY1 committed:
reset search button in searchcourseview when gotosearch is activated.

OliviaY1 committed:
Dynamic Result displayed on SearchCourseResultView

OliviaY1 committed:
add todo

sushidoggg reviewed (changes requested):
The implementation of the matching algorithm is very well done, but there are several ways to improve the application of the strategy design pattern:

- Introduce interfaces `MatchTutorAlgorithm` and `MatchStudentAlgorithm`.
- Develop classes `MatchByTutorRating` and `MatchByStudentRating` that implement the previouly mentioned interfaces.
- In the use case interactor's matching method, pass in instances of `MatchTutorAlgorithm` or `MatchStudentAlgorithm`, instead of holding one instance of `MatchingAlgorithm` (this also aligns better with the Single Responsibility Principle).

By doing so, a matching algorithm can implement both interfaces if necessary (e.g., `MatchByPersonalityTags`), reducing dependencies on the `MatchingAlgorithm` class.

OliviaY1 committed:
- Introduce MatchTutorAlgorithm and MatchStudentAlgorithm.
- Include two instances of MatchTutorAlgorithm and MatchStudentAlgorithm in Search Course Use Case
- reducing dependencies on the MatchingAlgorithm class.

OliviaY1 committed:
Merge branch 'main' into oliviaResult

OliviaY1 commented:
-  introduced MatchTutorAlgorithm and MatchStudentAlgorithm 
-  introduced two instances of MatchTutorAlgorithm and MatchStudentAlgorithm in SearchCourseUseCase

sushidoggg reviewed (approved):
Great job! The code nicely adheres to the SOLID Principles and uses an appropirate design pattern!
